ZeldaFullEditor/Gui/MainTabs/TextEditor.cs

```diff
@@ -711,690 +984,719 @@ private void TextListbox_SelectedIndexChanged(object sender, EventArgs e)
-                    {
-                        fileStream.Read(data, 0, 0x1000 + 100);
-                        fileStream.Close();
-                    }
-
-                    for (int i = 0; i < 0x1000; i++)
-                    {
-                        /*
+			// TODO: Need to adjust this because it keeps moving where the cursor is.
+			textBox1.Text = AddNewLinesToCommands(CurrentMessage.ContentsParsed);
+
+			DrawMessagePreview();
+
+			pictureBox1.Refresh();
+		}
+
+		public void DrawStringToPreview(string str)
+		{
+			foreach (char c in str)
+			{
+				DrawCharacterToPreview(c);
+			}
+		}
+
+		public void DrawCharacterToPreview(char c)
+		{
+			DrawCharacterToPreview(FindMatchingCharacter(c));
+		}
+
+		public void DrawCharacterToPreview(params byte[] text)
+		{
+			foreach (byte value in text)
+			{
+				if (skipNext)
+				{
+					skipNext = false;
+					continue;
+				}
+
+				if (value < NumberOfCharacters)
+				{
+					int srcy = value >> 4;
+					int srcx = value & 0xF;
+
+					if (textPos >= 170)
+					{
+						textPos = 0;
+						textLine++;
+					}
+
+					DrawTileToPreview(textPos, textLine * 16, srcx, srcy, 0, 1, 2);
+					textPos += widthArray[value];
+				}
+				else if (value == 0x73) // scroll text
+				{
+					textPos = 0;
+					textLine += 1;
+				}
+				else if (value == 0x74) // line 1
+				{
+					textPos = 0;
+					textLine = 0;
+				}
+				else if (value == 0x75) // line 2
+				{
+					textPos = 0;
+					textLine = 1;
+				}
+				else if (value == 0x76) // line 3
+				{
+					textPos = 0;
+					textLine = 2;
+				}
+				// command with parameters
+				else if (value == 0x6B || value == 0x6D || value == 0x6E || value == 0x77 || value == 0x78 || value == 0x79 || value == 0x7A)
+				{
+					skipNext = true;
+
+					continue;
+				}
+				else if (value == 0x6C) // BCD numbers
+				{
+					DrawCharacterToPreview('0'); // draw a big character since this command draws characters
+					skipNext = true;
+
+					continue;
+				}
+				else if (value == 0x6A) // player name
+				{
+					// Includes parentheses to be longer, since player names can be up to 6 characters.
+					DrawStringToPreview("(NAME)");
+				}
+				else if (value >= DictionaryBase && value < (DictionaryBase + NumberOfDictionaryEntries))
+				{
+					var dictionaryEntry = AllDictionaries.FirstOrDefault(dictionary => dictionary.RealID == value);
+					if (dictionaryEntry != null)
+					{
+						DrawCharacterToPreview(dictionaryEntry.Data);
+					}
+				}
+			}
+		}
+
+		public unsafe void DrawMessagePreview() // From Parsing.
+		{
+			textLine = 0;
+			// TODO try Span<byte> ?
+			byte* ptr = (byte*) GFX.currentfontgfx16Ptr.ToPointer();
+
+			for (int i = 0; i < (172 * 4096); i++)
+			{
+				ptr[i] = 0;
+			}
+
+			textPos = 0;
+			DrawCharacterToPreview(CurrentMessage.Data);
+
+			shownLines = 0;
+		}
+
+		public unsafe void DrawTileToPreview(int x, int y, int srcx, int srcy, int pal, int sizex = 1, int sizey = 1)
```

Comment:
Reading this now I can see why text rendering can be slow if we're copying the tile previews into a bitmap on the fly like this. Might be worth considering caching the data into a container so we can access already rendered letters.

Reply:
Done